File: js/taskbook_assignment.js
// Copyright 2026 FireCal LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

'use strict';

const { readDate, dateToIso } = require('./codec');
const { OrganizationSnapshot } = require('./organization_snapshot');
const { PersonSnapshot } = require('./person_snapshot');

/**
 * A person captured at the time of assignment, paired with the
 * timestamp of when the assignment was made. Reused for assignee
 * and evaluator slots.
 */
class AssignedPerson {
  constructor({ person, assignedAt = null }) {
    this.person = person;
    this.assignedAt = assignedAt;
    Object.freeze(this);
  }

  /** Reads the wire shape produced by toJSON(). */
  static fromJSON(m) {
    return new AssignedPerson({
      person: PersonSnapshot.fromJSON(m.person),
      assignedAt: readDate(m.assigned_at),
    });
  }

  /** Serializes to the snake-case wire shape (see `codec.js`). */
  toJSON() {
    const out = { person: this.person.toJSON() };
    if (this.assignedAt != null) out.assigned_at = dateToIso(this.assignedAt);
    return out;
  }
}

/**
 * An organization captured at the time of assignment, paired with
 * the timestamp of when the assignment was made. Used for the host
 * slot today and reusable for future organization-assignment slots.
 */
class AssignedOrganization {
  constructor({ organization, assignedAt = null }) {
    this.organization = organization;
    this.assignedAt = assignedAt;
    Object.freeze(this);
  }

  /** Reads the wire shape produced by toJSON(). */
  static fromJSON(m) {
    return new AssignedOrganization({
      organization: OrganizationSnapshot.fromJSON(m.organization),
      assignedAt: readDate(m.assigned_at),
    });
  }

  /** Serializes to the snake-case wire shape (see `codec.js`). */
  toJSON() {
    const out = { organization: this.organization.toJSON() };
    if (this.assignedAt != null) out.assigned_at = dateToIso(this.assignedAt);
    return out;
  }
}

/**
 * The assignment triple for a Taskbook: who is doing it, who will
 * evaluate it, and what organization is hosting it.
 */
class TaskbookAssignment {
  constructor({ assignee = null, evaluator = null, host = null } = {}) {
    this.assignee = assignee;
    this.evaluator = evaluator;
    this.host = host;
    Object.freeze(this);
  }

  /** Reads the wire shape produced by toJSON(). */
  static fromJSON(m) {
    return new TaskbookAssignment({
      assignee: m.assignee == null ? null : AssignedPerson.fromJSON(m.assignee),
      evaluator: m.evaluator == null ? null : AssignedPerson.fromJSON(m.evaluator),
      host: m.host == null ? null : AssignedOrganization.fromJSON(m.host),
    });
  }

  /** Serializes to the snake-case wire shape (see `codec.js`). */
  toJSON() {
    const out = {};
    if (this.assignee != null) out.assignee = this.assignee.toJSON();
    if (this.evaluator != null) out.evaluator = this.evaluator.toJSON();
    if (this.host != null) out.host = this.host.toJSON();
    return out;
  }
}

module.exports = {
  TaskbookAssignment,
  AssignedPerson,
  AssignedOrganization,
};

File: dart/lib/taskbook_assignment.dart
// Copyright 2026 FireCal LLC. Apache-2.0.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'codec.dart';
import 'organization_snapshot.dart';
import 'person_snapshot.dart';

/// The assignment triple for a Taskbook: who is doing it, who will
/// evaluate it, and what organization is hosting it.
class TaskbookAssignment {
  final AssignedPerson? assignee;
  final AssignedPerson? evaluator;
  final AssignedOrganization? host;

  const TaskbookAssignment({this.assignee, this.evaluator, this.host});

  /// Reads the wire shape produced by [toMap].
  factory TaskbookAssignment.fromMap(Map<String, dynamic> m) =>
      TaskbookAssignment(
        assignee: m['assignee'] == null
            ? null
            : AssignedPerson.fromMap(
                (m['assignee'] as Map).cast<String, dynamic>()),
        evaluator: m['evaluator'] == null
            ? null
            : AssignedPerson.fromMap(
                (m['evaluator'] as Map).cast<String, dynamic>()),
        host: m['host'] == null
            ? null
            : AssignedOrganization.fromMap(
                (m['host'] as Map).cast<String, dynamic>()),
      );

  /// Serializes to the snake-case wire shape (see `codec.dart`).
  Map<String, dynamic> toMap() => {
        if (assignee != null) 'assignee': assignee!.toMap(),
        if (evaluator != null) 'evaluator': evaluator!.toMap(),
        if (host != null) 'host': host!.toMap(),
      };
}

/// A person captured at the time of assignment, paired with the
/// timestamp of when the assignment was made. Reused for assignee
/// and evaluator slots.
class AssignedPerson {
  final PersonSnapshot person;
  final DateTime? assignedAt;

  const AssignedPerson({required this.person, this.assignedAt});

  /// Reads the wire shape produced by [toMap].
  factory AssignedPerson.fromMap(Map<String, dynamic> m) => AssignedPerson(
        person: PersonSnapshot.fromMap(
            (m['person'] as Map).cast<String, dynamic>()),
        assignedAt: readDateTime(m['assigned_at']),
      );

  /// Serializes to the snake-case wire shape (see `codec.dart`).
  Map<String, dynamic> toMap() => {
        'person': person.toMap(),
        if (assignedAt != null) 'assigned_at': assignedAt,
      };
}

/// An organization captured at the time of assignment, paired with
/// the timestamp of when the assignment was made. Used for the host
/// slot today and reusable for future organization-assignment slots.
class AssignedOrganization {
  final OrganizationSnapshot organization;
  final DateTime? assignedAt;

  const AssignedOrganization({required this.organization, this.assignedAt});

  /// Reads the wire shape produced by [toMap].
  factory AssignedOrganization.fromMap(Map<String, dynamic> m) =>
      AssignedOrganization(
        organization: OrganizationSnapshot.fromMap(
            (m['organization'] as Map).cast<String, dynamic>()),
        assignedAt: readDateTime(m['assigned_at']),
      );

  /// Serializes to the snake-case wire shape (see `codec.dart`).
  Map<String, dynamic> toMap() => {
        'organization': organization.toMap(),
        if (assignedAt != null) 'assigned_at': assignedAt,
      };
}
